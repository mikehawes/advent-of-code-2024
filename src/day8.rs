pub mod antenna_map;
