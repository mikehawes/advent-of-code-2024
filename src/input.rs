use std::fs::read_to_string;
use std::io;
use std::path::{Path, PathBuf};

pub fn input_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let full_path = file_in_src(path);
    read_to_string(full_path)
}

pub fn file_in_src<P: AsRef<Path>>(path: P) -> PathBuf {
    let mut full_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    full_path.push("src");
    full_path.push(path);
    full_path
}
