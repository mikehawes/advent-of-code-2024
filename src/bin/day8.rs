use advent_of_code_2024::day8::antenna_map::AntennaMap;
use advent_of_code_2024::input::input_to_string;

fn main() {
    let input = input_to_string("day8/input.txt").unwrap();
    let map = AntennaMap::parse(input.as_str());
    println!("Part 1: {}", map.count_unique_antinode_locations());
}
