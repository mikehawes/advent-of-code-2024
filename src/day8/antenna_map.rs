use std::collections::hash_map::Entry;
use std::collections::HashMap;

pub struct AntennaMap {
    width: usize,
    height: usize,
    positions_by_frequency: HashMap<char, Vec<Point>>,
}

pub(super) type Point = (usize, usize);

impl AntennaMap {
    pub fn parse(string: &str) -> AntennaMap {
        let mut width = 0;
        let mut height = 0;
        let mut positions_by_frequency = HashMap::new();
        for (y, line) in string.lines().enumerate() {
            width = line.len();
            height += 1;
            for (x, c) in line.chars().enumerate() {
                match c {
                    '.' => {}
                    _ => match positions_by_frequency.entry(c) {
                        Entry::Occupied(positions) => positions.into_mut(),
                        Entry::Vacant(entry) => entry.insert(vec![]),
                    }
                    .push((x, y)),
                }
            }
        }
        AntennaMap {
            width,
            height,
            positions_by_frequency,
        }
    }
    pub fn count_unique_antinode_locations(&self) -> usize {
        0
    }
}

#[cfg(test)]
mod tests {
    use crate::day8::antenna_map::{AntennaMap, Point};
    use crate::input::input_to_string;
    use insta::assert_snapshot;
    use std::collections::HashMap;

    #[test]
    fn can_plot_example() {
        let input = input_to_string("day8/example.txt").unwrap();
        assert_snapshot!(plot_map(&AntennaMap::parse(input.as_str())));
    }

    fn plot_map(map: &AntennaMap) -> String {
        let antenna_index = index_antenna_by_position(map);
        let mut str = "".to_string();
        for y in 0..map.height {
            for x in 0..map.width {
                str.push(*antenna_index.get(&(x, y)).unwrap_or(&'.'));
            }
            str.push('\n');
        }
        str
    }
    fn index_antenna_by_position(map: &AntennaMap) -> HashMap<Point, char> {
        map.positions_by_frequency
            .iter()
            .flat_map(|(frequency, positions)| {
                positions.iter().map(|position| (*position, *frequency))
            })
            .collect()
    }
}
