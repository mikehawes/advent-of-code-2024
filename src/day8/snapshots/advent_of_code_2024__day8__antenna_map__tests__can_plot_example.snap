---
source: src/day8/antenna_map.rs
assertion_line: 52
expression: "plot_map(&AntennaMap::parse(input.as_str()))"
snapshot_kind: text
---
............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............
